Choice parsing in toolchain

kshokhin opened the pull request:
Implement Choice parsing in toolchain according to [design](https://github.com/carbon-language/carbon-lang/blob/trunk/docs/design/sum_types.md).

kshokhin committed:
Build AST for Choice definitions

kshokhin committed:
Update choice alternative parsing

Use DeclNameAndParamsAsOptional state instead of manual nodes adding.

zygoloid commented on toolchain/parse/typed_nodes_test.cpp:
```diff
@@ -284,5 +284,84 @@ TEST_F(TypedNodeTest, CategoryMatches) {
 #include "toolchain/parse/node_kind.def"
 }
 
+TEST_F(TypedNodeTest, Choice) {
```
We don't generally expect unit tests for each kind of typed node. We check that the file tests (eg, testdata/choice/basic.carbon) can be successfully extracted as typed nodes as part of parse tree verification, so there is test coverage for extraction of these new typed nodes even without this added test.

zygoloid commented on toolchain/parse/typed_nodes.h:
```diff
@@ -677,6 +677,37 @@ struct IfExprElse {
+      NodeKind::ChoiceDefinitionStart.Define(NodeCategory::None);
+
+  ChoiceIntroducerId introducer;
+  AnyNameComponentId name;
```
We probably need a `llvm::SmallVector<AnyModifierId> modifiers;` here for things like `public choice X { ... }`.

zygoloid reviewed (commented):
The changes here look good to me and seem to be largely complete, and the tests are passing, but this PR has been in draft status for a week or so, so I wanted to check: are you looking for review / feedback, or are you still planning to do further work on this prior to asking for a review? I've left a couple of comments in case you want feedback (additionally: some negative tests for parse errors would be great), but generally I'm excited to see this getting worked on :) Thank you!

kshokhin commented:
> The changes here look good to me and seem to be largely complete, and the tests are passing, but this PR has been in draft status for a week or so, so I wanted to check: are you looking for review / feedback, or are you still planning to do further work on this prior to asking for a review? I've left a couple of comments in case you want feedback (additionally: some negative tests for parse errors would be great), but generally I'm excited to see this getting worked on :) Thank you!

Thanks for the feedback!

The PR is a draft mostly because there's almost no parse errors handling, it's still work to be done.

I have also a couple of questions and wanted to ask them in the discord, and waited for the access to the toolchain channel. However there's still no access for some reason.

The main purpose of making this PR as a draft was to gain the initial feedback on the overall approach and make sure that I'm not completely messing the things up :)

kshokhin committed:
Add handling of parse errors for choice header

kshokhin committed:
Add implicit parameters list to the choice signature

kshokhin committed:
Add alternative name parsing errors handling

kshokhin committed:
Add modifiers parsing

zygoloid commented on toolchain/parse/typed_nodes.h:
```diff
@@ -677,6 +677,39 @@ struct IfExprElse {
+
+  ChoiceDefinitionStartId signature;
+  struct Alternative {
+    AnyNameComponentId name;
```
I think we only want to allow simple identifier names here. I don't have a strong preference as to whether that's validated in the parser or in check, but either way, we should have a test covering this case, for example:
```
choice X { A.B }
```

zygoloid commented on toolchain/parse/handle_choice.cpp:
```diff
@@ -0,0 +1,97 @@
+
+    context.SkipPastLikelyEnd(*context.position());
+
+    state.state = State::ChoiceAlternativeFinish;
+    state.subtree_start = {};
+    state.token = *context.position();
+    state.has_error = true;
+    context.PushState(state);
```
I think we're not getting any benefit from reusing the previous state here, given that we overwrite all of its fields. I think it'd be a little more obvious what we're doing if we push a fresh state:
```suggestion
    context.PushState(State::ChoiceAlternativeFinish);
    context.ReturnErrorOnState();
```
(It might also be reasonable to move the `PushState(ChoiceAlternativeFinish)` calls from here and line 68 to before the `if (!context.PositionIs...`, given that we do that unconditionally.)

zygoloid reviewed (commented):
Thank you, this looks great! Just a couple of trivial things, then we can merge this.

kshokhin committed:
Refactor choice alternative parsing

kshokhin committed:
Update choice alternative parsing

1) Only simple identifiers are allowed
2) Treat implicit parameters list as an error

zygoloid reviewed (approved):
Awesome, thank you!
